bisection.go

```diff
@@ -57,7 +57,20 @@ func (b *Bisection) Init(loc LinesearchLocation, step float64, _ *FunctionInfo)
 }
 
 func (b *Bisection) Finished(loc LinesearchLocation) bool {
-	return StrongWolfeConditionsMet(loc.F, loc.Derivative, b.initF, b.initGrad, b.currStep, 0, b.GradConst)
+	// Don't finish the linesearch until a minimum is found that is better than
+	// the best point found so far. We want to end up in the lowest basin of
+	// attraction
+	minF := b.initF
+	if b.maxF < b.initF {
+		minF = b.maxF
+	}
+	if b.minF < minF {
+		minF = b.minF
+	}
+	if StrongWolfeConditionsMet(loc.F, loc.Derivative, minF, b.initGrad, b.currStep, 0, b.GradConst) {
```

Comment:
The problem that I was seeing was this:
````
\                  /
   \    /\       /
     \/   \    /
            \/             
````
The line search starts on the left, steps to a point on the right. This point on the right has a lower function value than the leftmost minimum, but has a high gradient and so does not meet the Wolfe conditions. The linesearch bisects, and ends up in the left minimum. The gradient is small, but the function value is higher than we had previously found. The goal of this change is to fix the behavior of Bisection such that the linesearch terminates in the rightmost minimum instead of the leftmost.

The Wolfe conditions are that there is a significant decrease in the function value (which for Bisection means that f_final <= f_start), and that there is a proportional decrease in the gradient |g_final| < alpha |g_start|. One must do the gradient comparison with reference to g_init, otherwise the Wolfe condition may not be met. The function criterion is that f_final <= f_start, but this is f_final <= f_min <= f_start, so this termination criteria still meets Wolfe.